### DISCUSSION TITLE
High-level + endian possible? | which way has lowest system load?

### DISCUSSION THREAD

**Discussion by gutschein** [aldas/modbus-tcp-client#80] [Q&A] [answered]
I've 2 questions: 

- I'd prefer the high-level API for my system, but I need another endian (BIG_ENDIAN_LOW_WORD_FIRST)). Is ist possible to change endian directly for each address. Maybe something like this from your example
`->int16(657, 'battery3_voltage_wo','BIG_ENDIAN_LOW_WORD_FIRST')`
or is it a must to use the low-level solution for another endian?

- If I'd like to read e.g. 20 addresses ~ 15sec. Which is the best way from the perspective of system load?

**Comment by aldas** [comment]
1. Currently no (i'll add todo issue for that). But you can use callback to achieve similar result

```php
    ->int16(657, 'battery3_voltage_wo', function ($result, $readRegisterAddress, $response) {
        return $response->getWordAt(657)->getInt16(\ModbusTcpClient\Utils\Endian::BIG_ENDIAN_LOW_WORD_FIRST)
    })
```

callback is called [after value is extracted](https://github.com/aldas/modbus-tcp-client/blob/0ce52f3a7ade81f34c470249ea7dc611b5f4359f/src/Composer/Read/Register/ReadRegisterAddress.php#L82) and it has access to response so you can overwrite value with new value.


2. for `If I'd like to read e.g. 20 addresses ~ 15sec. Which is the best way from the perspective of system load?` 

From request point of view couple of requests (I assume those 20 addresses are not that wide apart that many requests are split) does not create much load. 
I have a example for CLI poller that you could modify to fire every 15 sec and do request + store results to somewhere
https://github.com/aldas/modbus-tcp-client/blob/master/examples/example_cli_poller.php

**Reply by gutschein** [reply]
Thank you Martti.
to 2. sounds good - I will try.
to 1.)... maybe I'm just not as good with that code stuff as I'd like to. I get errors. Maybe as my addresses are all `int32, uint32, uint64`? I Tried `->getUInt32`, but it doesn't work. The library contains maybe `getInt16 `and `getUInt16 `only by now?
And what is the meaning of the value 657 in `getWordArt(657)`? It seems to be a length or offset; but how do I derive the proper value?
```
  ->uint32(30845,'battery status',function ($result, $readRegisterAddress, $response) {
        return $response->getWordAt(657)->getUInt32(\ModbusTcpClient\Utils\Endian::BIG_ENDIAN_LOW_WORD_FIRST);
    })
```
`PHP Fatal error:  Uncaught ModbusTcpClient\Exception\InvalidArgumentException: offset out of bounds in /usr/local/edomi/main/include/php/modbus-tcp-client/vendor/aldas/modbus-tcp-client/src/Packet/ModbusFunction/ReadHoldingRegistersResponse.php:152
`

**Comment by aldas** [comment, marked as answer]
sorry, I did not try it out. `$readRegisterAddress` has field `address` and you can you this for extraction

```
 ->uint32(30845,'battery status',function ($result, $readRegisterAddress, $response) {
        return $response->getWordAt($readRegisterAddress->address)->getUInt32(\ModbusTcpClient\Utils\Endian::BIG_ENDIAN_LOW_WORD_FIRST);
    })
```

NB: this is on top of my head. I did not try it out :)

**Reply by aldas** [reply]
This is how extraction works https://github.com/aldas/modbus-tcp-client/blob/0ce52f3a7ade81f34c470249ea7dc611b5f4359f/src/Composer/Read/Register/ReadRegisterAddress.php#L48 which value we are trying to overwrite with out callback

**Reply by gutschein** [reply]
Don't worry, I t'll try it for you... :)   I really appreciate your quick respsonse!
But it doesn't work with another error:
```
  ->uint32(30845,'battery status', function ($result, $readRegisterAddress, $response) {
        return $response->getWordAt($readRegisterAddress->address)->getUInt32(\ModbusTcpClient\Utils\Endian::BIG_ENDIAN_LOW_WORD_FIRST);
    })
```
`PHP Fatal error:  Uncaught Error: Cannot access protected property ModbusTcpClient\Composer\Read\Register\ReadRegisterAddress::$address in /usr/local/edomi/main/include/php/modbus-tcp-client/test.php:11`

**Reply by aldas** [reply]
Replace `$readRegisterAddress->address` with `$readRegisterAddress->getAddress()`. Base class allow us to access address with [getAddress](https://github.com/aldas/modbus-tcp-client/blob/0ce52f3a7ade81f34c470249ea7dc611b5f4359f/src/Composer/RegisterAddress.php#L26) method

**Reply by gutschein** [reply]
seems to work...Address is accessable... :)
next issue: 
`PHP Fatal error:  Uncaught Error: Call to undefined method ModbusTcpClient\Packet\Word::getUInt32() in /usr/local/edomi/main/include/php/modbus-tcp-client/test.php:11`

**Reply by aldas** [reply]
You are currently trying to extract uint32 from word/register. 1 register/word is 16 bit.  For 32bit values you need to work with double words.  So replace `getWordAt` with `getDoubleWordAt`

```
  ->uint32(30845,'battery status', function ($result, $readRegisterAddress, $response) {
        return $response->getDoubleWordAt($readRegisterAddress->getAddress())->getUInt32(\ModbusTcpClient\Utils\Endian::BIG_ENDIAN_LOW_WORD_FIRST);
    })
```

**Reply by aldas** [reply]
NB: If all your values are with same endian you could just change global/default endian value

add somewhere at the beggining of your code:
```
Endian::$defaultEndian = Endian::BIG_ENDIAN_LOW_WORD_FIRST;
```

**Reply by gutschein** [reply]
Perfect! You really impress me. Thank you. SOLVED! 

```
            [battery status endian] => 27
            [battery status pure] => 1769472
```
with
```
  ->uint32(30845,'battery status endian', function ($result, $readRegisterAddress, $response) {
        return $response->getDoubleWordAt($readRegisterAddress->getAddress())->getUInt32(\ModbusTcpClient\Utils\Endian::BIG_ENDIAN);
    })
  ->uint32(30845,'battery status pure')
```

I will try to change the default endian, too.

**Reply by gutschein** [reply]
With a leading 
`use ModbusTcpClient\Utils\Endian;`
the change of the default value works perfect! That make's it much easier and the code much lean. 👍 
This info is less for you...more for people like me with the same issues... :)
